packages/tpetra/core/test/MultiVector/Transform.cpp

```diff
@@ -138,6 +139,8 @@ namespace { // (anonymous)
     }
 
     Y.putScalar (418.0);
+    Kokkos::fence();
```

Comment:
It's interesting that we needed a fence here.  The last `transform` call should have marked the MultiVector as modified on device.  This means that `putScalar` should run on device. The `transform` that follows should notice that `Y` has been modified on device, and call `sync_host`.  It used to be the case that `Kokkos::DualView::sync_host` called `Kokkos::fence` even if the two memory spaces were the same.  Now it does nothing.  Surprise! ;-)

I would say, rather than putting a fence here, try changing the implementation of `Tpetra::MultiVector::sync_host` to look like this:
```cpp
if (need_sync_host ()) {
  execution_space ().fence ();
  view_.sync_host ();
}
```
That should fix this issue.  If it doesn't, then there's something wrong.

Reply:
@mhoemmen Thanks for the feedback. I verified putScalar was acting on device. Tried your recommended change but then it seems transform never actually calls sync_host or sync_device. You are expecting transform to do this, so perhaps we want to call it at the beginning of the transform for the requested exec space?

In this case transform calls transform_mv_notSameObject and I observe we'll need a fence before the parallel_for, shown below. But I expect we resolve this earlier, probably at the beginning of the transform.

```
        withLocalAccess
          ([=] (const input_view_type& input_lcl,
            using functor_type = MultiVectorUnaryTransformLoopBody<
            input_view_type, output_view_type, UnaryFunctionType, LO>;
            functor_type g (input_lcl, output_lcl, f);

            const LO lclNumRows = static_cast<LO> (input_lcl.extent (0));
            using range_type = Kokkos::RangePolicy<ExecutionSpace, LO>;
            range_type range (execSpace, 0, lclNumRows);
// we need the fence before this parallel_for
            Kokkos::parallel_for (kernelLabel, range, g);
          },
          readOnly (input).on (memSpace),
          writeOnly (output).on (memSpace));
```
Also it seems I made a mistake on two other fences - I could not reproduce their necessity. Only the putScalar calls before the transform for host execution space are necessary. The transform on device does not a fence since the prior putScalar was also on device.

Reply:
`withLocalAccess` should call the templated form of `Tpetra::MultiVector::sync`.  See lines 118-151 of [`Tpetra_withLocalAccess_MultiVector.hpp`](https://github.com/trilinos/Trilinos/blob/5bd5b68a3284b1bfacdafebcfd23484057fdf486/packages/tpetra/core/src/Tpetra_withLocalAccess_MultiVector.hpp#L143).  If it's not, then that might indicate that `Tpetra::MultiVector::need_sync` is not doing the right thing.

Reply:
As a temporary test, could you please try the following change to that file?  Replace lines 142-144 with the following:
```cpp
if (LA.G_.template need_sync<space>()) {
  if (is_host) {
    using dev_exec_space = typename dual_view_type::t_dev::execution_space;
    dev_exec_space().fence();
  }
  LA.G_.template sync<space>();
}
```
Alternately, we could just change `Tpetra::MultiVector::sync` always to do a fence, as above.

Reply:
You had asked me to originally modify need_sync_host and need_sync_device, but I'm realizing now we also have need_sync which is called in this case. Do we want to modify the need_sync the same way, with a fence on the opposite of host or device depending on TargetDeviceType. 

```
    //! Whether this MultiVector needs synchronization to the given space.
    template<class TargetDeviceType>
    bool need_sync () const {
// I think we want to fence here similar to your original suggestions
// fence host if TargetDeviceType is device
// fence device if TargetDeviceType is host
      return view_.template need_sync<TargetDeviceType> ();
    }

    //! Whether this MultiVector needs synchronization to the host.
    bool need_sync_host () const;

    //! Whether this MultiVector needs synchronization to the device.
    bool need_sync_device () const;
```

Reply:
@MicheldeMessieres Wait, I didn't ask you to modify `need_sync*`.  That doesn't make sense -- the point of `need_sync*` is to check the flags, not to do work.

Here's what should actually happen:

1. `sync_host` should only fence if a fence is needed, i.e., if `need_sync_host()` is true.
2. `sync<Space>` should only fence if a fence is needed (i.e., if `need_sync<Space>()` is true), and only if sync'ing to host.

I'm all for getting this stuff over with, so whatever, just put the fences wherever, but please just put a big comment over them saying something like "THIS IS WEIRD WE SHOULDN'T NEED THIS SEE `#6617`" and I'll go back to it later.  I don't want to waste too much time when it could be some issue with DualView or even with `Kokkos::fence()` being a lie.

Reply:
Sorry - I meant you asked me to modify sync_host and sync_device and then noticed we also had the templated sync in the decl. Then did my own brain transform and edited the wrong thing.

I updated the PR to remove fences from Transform.cpp and added fences with a TODO comment in the transform calls. This works but I need to investigate further.

The user choice of host (execSpace) is not passed into withLocalAccess so I don't think it can know about the choice and act on it. The problem is when host is requested in the transform call and data is on device (or vice versa I expect). I will educate myself on exactly what these range policies are doing and learn more about this setup. Then the proper design may be more apparent.